### PR TITLE
Steps towards reproducible builds

### LINKED ISSUE #24473: Rustdoc output slightly nondeterministic
There's a bit of nondeterminism in rustdoc output across runs even when the source doesn't change: https://github.com/sfackler/rust-openssl/commit/1cb83ffe9b974980704b7e809aa7279cf02a7252 I'm guessing it's due to randomized HashMap iteration order.

It'd be nice to switch to a non keyed Hasher or maybe just TreeMaps in rustdoc to avoid "phantom" documentation updates.

### PR DESCRIPTION
**Pull request by jonas-schievink**
cc #34902

Running `make dist` twice will result in a rustc tarball where only `librustc_back.so`, `librustc_llvm.so` and `librustc_trans.so` differ. Building `libstd` and `libcore` twice with the same compiler and flags produces identical artifacts.

The third commit should close #24473

### CODE REVIEW

File: src/librustc/infer/freshen.rs [hunk 1]
```diff
@@ -32,15 +32,16 @@
...
+use util::nodemap::FnvHashMap;
+use std::collections::hash_map::Entry;
 
 use super::InferCtxt;
 use super::unify_key::ToType;
 
 pub struct TypeFreshener<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
     infcx: &'a InferCtxt<'a, 'gcx, 'tcx>,
     freshen_count: u32,
-    freshen_map: hash_map::HashMap<ty::InferTy, Ty<'tcx>>,
```
Comment (arielb1) [inline src/librustc/infer/freshen.rs]: That sounds like a performance problem.

Comment (jonas-schievink) [inline src/librustc/infer/freshen.rs]: I was surprised to see this, too. I'll try to do a little benchmark to see if this patch improves type checking performance.

Comment (jonas-schievink) [inline src/librustc/infer/freshen.rs]: Some tests on `syntex_syntax`:

With this PR:
```
time: 0.557; rss: 230MB item-types checking
time: 5.720; rss: 262MB item-bodies checking

time: 0.574; rss: 230MB item-types checking
time: 5.779; rss: 262MB item-bodies checking

time: 0.559; rss: 231MB item-types checking
time: 5.761; rss: 261MB item-bodies checking

time: 0.566; rss: 232MB item-types checking
time: 5.708; rss: 263MB item-bodies checking
```

Without this PR (current master):
```
time: 0.560; rss: 232MB item-types checking
time: 5.768; rss: 263MB item-bodies checking

time: 0.565; rss: 230MB item-types checking
time: 5.907; rss: 261MB item-bodies checking

time: 0.561; rss: 230MB item-types checking
time: 5.810; rss: 261MB item-bodies checking

time: 0.557; rss: 231MB item-types checking
time: 5.812; rss: 261MB item-bodies checking
```

Comment (arielb1) [inline src/librustc/infer/freshen.rs]: nice!

**Comment by jonas-schievink** [comment]
Works every time :stuck_out_tongue_winking_eye:

**Comment by jonas-schievink** [comment]
```
thread '[incremental] incremental\ich_method_call_trait_scope.rs' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 3, message: "The system cannot find the path specified." } }', C:\bot\slave\auto-win-gnu-32-opt-rustbuild\build\src\libcore\result.rs:783
```

New spurious failure?

**Comment by jonathandturner** [comment]
@jonas-schievink - looks like it, hopefully it's fixed now.  I'm going to be optimistic and retry

@bors retry

**Comment by jonas-schievink** [comment]
Apparently it doesn't really work on MSVC (no idea what's up). I've reverted the test changes in order to at least get *something*.

**Pull request merged**